src/types/any.rs

```diff
@@ -83,10 +83,24 @@ impl PyAny {
     where
         N: IntoPy<Py<PyString>>,
     {
-        let py = self.py();
-        let attr_name = attr_name.into_py(py);
+        // PyObject_HasAttr suppresses all exceptions, which was the behaviour of `hasattr` in Python 2.
+        // Use an implementation which suppresses only AttributeError, which is consistent with `hasattr` in Python 3.
+        self.hasattr_impl(attr_name.into_py(self.py()))
+    }
 
-        unsafe { Ok(ffi::PyObject_HasAttr(self.as_ptr(), attr_name.as_ptr()) != 0) }
+    fn hasattr_impl(&self, attr_name: Py<PyString>) -> PyResult<bool> {
+        let py = self.py();
+        let result = unsafe {
+            PyObject::from_owned_ptr_or_err(
+                py,
+                ffi::PyObject_GetItem(self.as_ptr(), attr_name.as_ptr()),
+            )
+        };
+        match result {
+            Ok(_) => Ok(true),
```

Comment:
This does imply an unnecessary increment-then-decrement of the reference count, doesn't it? Also why `GetItem` instead of `GetAttr` as suggested by [the docs](https://docs.python.org/3/c-api/object.html#c.PyObject_HasAttr)?

If we do eat the reference count modifications, couldn't this just call `getattr(name)` and match on the error?

Reply:
> This does imply an unnecessary increment-then-decrement of the reference count, doesn't it?

Sadly yes, though this was always the case, just abstracted away in `PyObject_HasAttr`.

> Also why `GetItem` instead of `GetAttr` as suggested by [the docs](https://docs.python.org/3/c-api/object.html#c.PyObject_HasAttr)?

A bug, should be `GetAttr`, this is exactly why I need to push a test 😄

Reply:
Changed the implementation to go via `getattr` as suggested (and pushed tests).

Reply:
Just to make sure I understand it correctly: Calling `_getattr` instead of `getattr` avoids the `into_ref(py)` which involves the pool? So when we drop the pool and `getattr` returns `PyAny<'py>` this distinction can be removed?

Reply:
Yes that's exactly right. Avoiding the pool means that we don't end up holding a reference to whatever we checked for in `hasattr` in the pool - so it's just a micro-optimization for now that should reduce away soon.